src/Confluent.Kafka/Impl/SafeKafkaHandle.cs

```diff
@@ -429,6 +477,24 @@ internal bool ConsumerPoll(out Message message, IntPtr millisecondsTimeout)
 
             long timestamp = LibRdKafka.message_timestamp(msgPtr, out IntPtr timestampType);
 
+            Headers headers = new Headers();
```

Comment:
I didn't know about that procedure. I haven't benchmarked this. In the case there are no headers, I think perf is fine. In the interests of getting this PR in, I think I should put a TODO/make a task and consider this an enhancement.

Reply:
That's fine, but please do a before-and-after performance benchmark to see what the actual cost is, prior to merging the PR.

Reply:
thanks for the support @gccmbr :-)

Reply:
A problem with doing the on-demand marshaling as you suggest is that the `Message` and `Headers` classes now need to be `IDisposable`, which is potentially very awkward for the end user to use correctly (need to wrap in `try/catch`, using `using` etc. everywhere in order to make sure dispose is called) This impacts even people not using headers since the `Message` class needs to be `IDisposable` as it manages the headers object. In practice the user could ignore this in that case and get away with it, but that's not the point of the contract - it all acts to add confusion. This sort of thing is/was a problem with `Windows.Forms` which wraps native components.

I'm strongly in favor of pre-emptively marshaling unless there is a huge perf cost (yes, will benchmark).